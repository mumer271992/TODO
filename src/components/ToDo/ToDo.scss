.todo-item{
  > div:first-child {
    width: 40%;
  }
  @media (max-width: 675px) {
    flex-direction: column !important;
  }
}