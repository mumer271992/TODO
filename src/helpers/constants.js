export default {
  actionTypes: {
    create: 'CREATE',
    update: 'UPDATE',
  },
  status: {
    pending: 'pending',
    completed: 'completed'
  }
}