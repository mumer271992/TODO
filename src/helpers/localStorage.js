export const  create = (item) => {
  let dataList = getList();
  dataList.push(item);
  localStorage.setItem('todoData', JSON.stringify(dataList));
} 

export const getList = () => {
  let data = localStorage.getItem('todoData');
  return data !== null ? JSON.parse(data) : [];
}

export const update = (id, data) => {
  let dataList = getList();
  let targetIndex = dataList.findIndex(item => item.uuid === id);
  if (targetIndex > -1) {
      dataList[targetIndex] = { ...dataList[targetIndex], ...data };
  }
  localStorage.setItem('todoData', JSON.stringify(dataList));
}