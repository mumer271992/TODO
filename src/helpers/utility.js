/* All utility functions should be here */

export const formatDate = () => {
  const date = new Date();
  return `${date.getFullYear()}-${date.getMonth() + 1}-${date.getDate()}`;
}