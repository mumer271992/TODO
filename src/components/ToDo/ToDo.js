import React from 'react';

import { formatDate } from '../../helpers/utility';
import constants from '../../helpers/constants';
import './ToDo.scss';

class ToDo extends React.Component {
  constructor(props){
    super(props);
    this.state = {
      status: '',
    }
    this.editHandler = this.editHandler.bind(this);
    this.onDelete = this.onDelete.bind(this);
    this.toggleStatus = this.toggleStatus.bind(this);
    this.restoreHandler = this.restoreHandler.bind(this);
  }

  editHandler() {
    const { todo, history } = this.props;
    history.push(`/todo/${todo.uuid}`);
  }

  onDelete() {
    const { updateTodo, todo } = this.props;
    const item = {
      deleted_at: formatDate(),
    };
    updateTodo(todo.uuid, item);
  }

  restoreHandler() {
    const { updateTodo, todo } = this.props;
    const item = {
      deleted_at: null,
    };
    updateTodo(todo.uuid, item);
  }

  toggleStatus() {
    const { updateTodo, todo } = this.props;
    let status = todo.status;
    let completed_at = null;
    if (status === constants.status.pending) {
      status = constants.status.completed;
      completed_at = formatDate();
    } else {
      status = constants.status.pending;
      completed_at = null;
    }
    const item = {
      status,
      completed_at,
    };
    updateTodo(todo.uuid, item);
  }
  

  render() {
    const { todo } = this.props;
    return (
      <li
        data-test="list-group-item"
        className="todo-item list-group-item d-flex flex-row justify-content-between"
      >
        <div data-test="todo-item-text">{ todo.text }</div>
        <div>
          { todo.due_date }
        </div>
        <div>
          {
            !todo.deleted_at ? (
              <div className="btn-group" role="group" aria-label="Basic example">
                <button type="button" data-test="todo-edit-btn" className="btn btn-primary " onClick={this.editHandler}>Edit</button>
            <button type="button" data-test="todo-mark-btn" className="btn btn-secondary" onClick={this.toggleStatus}>Mark {todo.status === 'pending' ? 'complete' : 'pending'}</button>
                <button type="button" data-test="todo-delete-btn" className="btn btn-danger" onClick={this.onDelete}>Delete</button>
              </div>
            ) : (
              <button type="button" data-test="todo-restore-btn" className="btn btn-secondary" onClick={this.restoreHandler}>Restore</button>
            )
          }
        </div>
      </li>
    );
  }
}

export default ToDo;
