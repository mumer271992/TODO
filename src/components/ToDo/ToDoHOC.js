import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

import ToDo from './ToDo';
import { updateTodo } from '../../actions/todos';

const mapDispatchToProps = dispatch => ({
  updateTodo: (uuid, todo) => dispatch(updateTodo(uuid, todo)),
});

export default withRouter(connect(null, mapDispatchToProps)(ToDo));
