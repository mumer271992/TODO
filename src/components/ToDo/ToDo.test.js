import React from 'react';
import { shallow } from 'enzyme';

import ToDo from './ToDo';
import { formatDate } from '../../helpers/utility';
import constants from '../../helpers/constants';

const updateTodo = jest.fn();

function shallowSetup(isDeleted = false) {
  const props = {
    updateTodo,
    todo: {
      id: Date.now(),
      text: 'This is todo item component test',
      status: constants.status.pending,
      due_date: formatDate(),
      completed_at: null,
      deleted_at: !isDeleted ? null : formatDate(),
    }
  }
  const wrapper = shallow(<ToDo {...props}/>);

  return {
    props,
    wrapper,
  }
}

describe('Active todo item component unit tests.', () => {
  it('Component should render successfully', () => {
    const shallowWrapper = shallowSetup();
    expect(shallowWrapper.wrapper.length).toEqual(1);
    expect(shallowWrapper.wrapper).toMatchSnapshot();
  });

  it('Active component should render with todo details', () => {
    const shallowWrapper = shallowSetup();
    const textWrapper = shallowWrapper.wrapper.find('[data-test="todo-item-text"]');
    const editBtnWrapper = shallowWrapper.wrapper.find('[data-test="todo-edit-btn"]');
    const deleteBtnWrapper = shallowWrapper.wrapper.find('[data-test="todo-delete-btn"]');
    const markBtnWrapper = shallowWrapper.wrapper.find('[data-test="todo-mark-btn"]');

    expect(textWrapper.text()).toEqual(shallowWrapper.props.todo.text);
    expect(editBtnWrapper.length).toEqual(1);
    expect(deleteBtnWrapper.length).toEqual(1);
    expect(markBtnWrapper.length).toEqual(1);
  });

  it('Clicking edit button should call', () => {
    const spy = jest.spyOn(ToDo.prototype, 'editHandler').mockImplementation(jest.fn());
    const shallowWrapper = shallowSetup();
    const editBtnWrapper = shallowWrapper.wrapper.find('[data-test="todo-edit-btn"]').first();
    editBtnWrapper.simulate('click');
    expect(spy).toHaveBeenCalled();
  });

  it('Clicking delete button should call', () => {
    const spy = jest.spyOn(ToDo.prototype, 'onDelete');
    const shallowWrapper = shallowSetup();
    const deleteBtnWrapper = shallowWrapper.wrapper.find('[data-test="todo-delete-btn"]').first();
    deleteBtnWrapper.simulate('click');
    expect(spy).toHaveBeenCalled();
  });

  it('Clicking mark button should call', () => {
    const spy = jest.spyOn(ToDo.prototype, 'toggleStatus');
    const shallowWrapper = shallowSetup();
    const markBtnWrapper = shallowWrapper.wrapper.find('[data-test="todo-mark-btn"]').first();
    markBtnWrapper.simulate('click');
    expect(spy).toHaveBeenCalled();
  });

  // /* The following test case are for delete todos where we show option to restore todo item. */

  it('Deleted component should render with todo details', () => {
    const shallowWrapper = shallowSetup(true);
    const textWrapper = shallowWrapper.wrapper.find('[data-test="todo-item-text"]');
    const restoreBtnWrapper = shallowWrapper.wrapper.find('[data-test="todo-restore-btn"]');

    expect(textWrapper.text()).toEqual(shallowWrapper.props.todo.text);
    expect(restoreBtnWrapper.length).toEqual(1);
  });

  it('Clicking restore button should call', () => {
    const spy = jest.spyOn(ToDo.prototype, 'restoreHandler');
    const shallowWrapper = shallowSetup(true);
    const restoreBtnWrapper = shallowWrapper.wrapper.find('[data-test="todo-restore-btn"]').first();
    restoreBtnWrapper.simulate('click');
    expect(spy).toHaveBeenCalled();
  });

});
