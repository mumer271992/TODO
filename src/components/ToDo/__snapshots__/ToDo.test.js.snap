// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Active todo item component unit tests. Component should render successfully 1`] = `ShallowWrapper {}`;
